src/main/java/org/breedinginsight/brapi/v2/BrAPIObservationsController.java

```diff
@@ -93,8 +106,96 @@ public HttpResponse observationsGet(@PathVariable("programId") UUID programId,
                                         @Nullable @QueryValue("externalReferenceSource") String externalReferenceSource,
                                         @Nullable @QueryValue("page") Integer page,
                                         @Nullable @QueryValue("pageSize") Integer pageSize) {
-        //TODO
-        return HttpResponse.notFound();
+        log.debug("observationsGet: fetching observations by filters");
+        Optional<Program> program = programService.getById(programId);
+        if(program.isEmpty()) {
+            log.warn("Program id: " + programId + " not found");
+            return HttpResponse.notFound();
+        }
+
+        try {
+
+            // TODO: BI-2506 - implement support for all query parameters.
+            List<Object> unsupportedParams = Stream.of(
+                    observationDbId,
+                    observationUnitDbId,
+                    observationVariableDbId,
+                    locationDbId,
+                    seasonDbId,
+                    observationTimeStampRangeStart,
+                    observationTimeStampRangeEnd,
+                    observationUnitLevelName,
+                    observationUnitLevelOrder,
+                    observationUnitLevelCode,
+                    observationUnitLevelRelationshipName,
+                    observationUnitLevelRelationshipOrder,
+                    observationUnitLevelRelationshipCode,
+                    observationUnitLevelRelationshipDbId,
+                    commonCropName,
+                    programDbId,
+                    trialDbId,
+                    germplasmDbId,
+                    externalReferenceID,
+                    externalReferenceId,
+                    externalReferenceSource
+            ).filter(Objects::nonNull).collect(Collectors.toList());
+
+            if (!unsupportedParams.isEmpty()) {
+                return HttpResponse.status(HttpStatus.NOT_IMPLEMENTED).body(
+                        new BrAPIObservationListResponse().metadata(new BrAPIMetadata().status(List.of(new BrAPIStatus().messageType(BrAPIStatus.MessageTypeEnum.ERROR)
+                                .message("Unsupported query parameter. Only studyDbId, page, and pageSize are supported."))))
+                );
+            }
+
+            // Get a filtered list of observations.
+            List<BrAPIObservation> observations = observationDAO.getObservationsByFilters(program.get(), studyDbId);
+
+            // Total number of records in the unpaged super set.
+            int totalCount = observations.size();
```

Comment:
We may want to consider making a generic version of this paging logic somewhere if we're going to be doing this again in the future but I'm fine putting it off until that point.